### DISCUSSION TITLE
Unable to calculate update time of a react native component using Sentry.withProfiler and using custom instrumentation

### DISCUSSION THREAD

**Discussion by IMniteshKumar** [getsentry/support#37] [Support] [answered]
We are using sentry in our React Native application ,  we want to capture the update time of our component  

**Button component is used from React-native-paper**

**counter.tsx file** 
```
function transactionFunc(name: string) {
  const transaction = Sentry.getCurrentHub().getScope()?.getTransaction() || Sentry.startTransaction({ name });
  return transaction;
}

const Counter: React.FunctionComponent = function () {
  const [counter, setCounter] = useState(0);

  useEffect(() => {
    const transaction = transactionFunc('counter-update-transaction');
    const span = Sentry?.getCurrentHub()?.getScope()?.getSpan();
    span?.finish();
  }, [counter]);

  useEffect(() => {
    return () => {
      const transaction = transactionFunc('counter-update-transaction');
      transaction?.finish();
    };
  }, []);

  const handlePress = () => {
    setCounter(counter + 1);
    const transaction = transactionFunc('counter-update-transaction');
    Sentry.getCurrentHub().configureScope((scope) => scope.setSpan(transaction));
    const span = transaction?.startChild({
      op: `onClick-update`,
      description: 'Update happens when we click button',
    });
  };

  return (
    <>
      <Button onPress={handlePress}>Increment</Button>
      <Text>{counter}</Text>
    </>
  ); }
  export default Sentry.withProfiler(Counter)
```
  
  **Sentry initialization file** 
  
```
  export const routingInstrumentation = new Sentry.ReactNavigationInstrumentation();

export const InitializeMetricsCodePushWrapper = () => {
  CodePush.getUpdateMetadata().then((update) => {
    if (update) {
      InitializeMetricsWhenCodePush(update.label);
    } else {
      InitializeMetrics();
    }
  });
};

const InitializeMetricsWhenCodePush = (codePushVersion: string) => {
  const release = DeviceInfo.getVersion() + '@codepush-' + codePushVersion;
  const dist = codePushVersion as string;
  Sentry.init({
    dsn: 'dsn from sentry',
    tracesSampleRate: 1.0,
    debug: true,
    release: release,
    dist: dist,
    environment: env.ENV,
    attachStacktrace: true,
    integrations: [
      new Sentry.ReactNativeTracing({
        routingInstrumentation,
      }),
    ],
  });
  SetTag('OTA_VERSION', env.OTA_VERSION);
};

const InitializeMetrics = () => {
  Sentry.init({
    dsn: 'dsn from sentry',
    tracesSampleRate: 1.0,
    debug: true,
    environment: env.ENV,
    integrations: [
      new Sentry.ReactNativeTracing({
        routingInstrumentation,
        tracingOrigins: ['localhost', /^\//, /^http:\/\//],
      }),
    ],
  });
};
```
**packages used for navigation** 
`
 "@react-navigation/native": "^5.2.1",
 "@react-navigation/stack": "^5.2.16",
`
**in the navigation file we followed this** 

```

const App = () => {
  const navigation = React.useRef();

  return (
    <NavigationContainer
      ref={navigation}
      onReady={() => {
        // Register the navigation container with the instrumentation inside onReady
        routingInstrumentation.registerNavigationContainer(navigation);
      }}
    >
      ...
    </NavigationContainer>
  );
};
```

Initially we tried using Sentry.withProfiler with  options { includeRender: true, includeUpdates: true }
but we were only getting the mount but not the component update time , 

if anyone could help us with this issue at the earliest that would be great.
Thanks

**Comment by AbhiPrasad** [comment, marked as answer]
Which component are you tracking with `Sentry.withProfiler`?

**Reply by IMniteshKumar** [reply]
Counter

**Reply by IMniteshKumar** [reply]
@AbhiPrasad  any resolution for the above query

**Reply by AbhiPrasad** [reply]
The Counter component won’t get any update spans attached to it because it doesn’t have any incoming props - just internal state. The React Profiler HOC relies on looking at changes to incoming props to determine an update. We recommend extracting out the inner component into another component that you can wrap so that the HOC can create update spans based on prop changes.

**Reply by IMniteshKumar** [reply]
![Screenshot 2022-06-28 at 2 49 33 PM](https://user-images.githubusercontent.com/107825927/176143050-e1d69e7c-e897-473c-a85b-da7ff09aa4ab.png)
Done that as well , it is showing the update time as 0.00 ms

**Reply by AbhiPrasad** [reply]
This is because the time it took to update was less than 0.01 ms, which is the max granularity we use, so it shows up as 0.00ms.

**Reply by IMniteshKumar** [reply]
<img width="1114" alt="Screenshot 2022-06-30 at 2 30 36 PM" src="https://user-images.githubusercontent.com/107825927/176637242-fce3e62f-32b0-45d2-a401-626dfe9fac55.png">
I have increase the update time now but still it is showing the same , as u can see in the above image

**Reply by AbhiPrasad** [reply]
Ah yeah, it seems I was mistaken. The update span is always a 0 duration span, it just represents when an update happens: https://github.com/getsentry/sentry-javascript/blob/b0efcb886050befbca8967fb531946c7bd643973/packages/react/src/profiler.tsx#L78

**Reply by IMniteshKumar** [reply]
is there a way to capture it  ?

**Reply by AbhiPrasad** [reply]
You can create your own instrumentation, (just like how it works above), and track the time between `shouldComponentUpdate`, and `componentDidUpdate`.

**Reply by IMniteshKumar** [reply]
cool , Just one thing does using  `Sentry.withProfiler` excessively  Cause any Performance issue.

**Reply by AbhiPrasad** [reply]
Yes, there is a possibility - but you should be fine for the most part. This is why we recommend not using it with components that rerender many times (like a textfield component for ex.)

**Reply by IMniteshKumar** [reply]
Is there a way to disable Sentry profiler from Sentry UI / remotely

**Reply by AbhiPrasad** [reply]
At the current moment - no, you’ll have to redeploy to make the change.

We are working on improving this - stay tuned!
